\documentclass[times, 11pt, a4paper]{article}
\usepackage{authblk}
\usepackage{lineno}
\linenumbers

\usepackage[utf8]{inputenc} %unicode support
\usepackage{amsmath}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{todonotes}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{amssymb}
\DeclareCaptionLabelFormat{r-parens}{\textbf{#2}}
\captionsetup[subfigure]{font={bf,small}, skip=1pt, singlelinecheck=false, labelformat=r-parens}
\usepackage{graphicx}
\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage{hyperref}
\renewcommand{\refname}{REFERENCES}
\renewcommand\Authfont{\fontsize{10}{11}\selectfont}
\renewcommand\Affilfont{\fontsize{9}{11}\selectfont}
\usepackage{spverbatim}
\begin{document}

\providecommand{\keywords}[1]{\textbf{Keywords:} #1}

\title{\Large Mapping of RNA modifications by direct Nanopore sequencing and JACUSA2}

\date{}
\author[1,2,3]{Christoph Dieterich\thanks{christoph.dieterich@uni-heidelberg.de}}
\author[1,2]{Amina Lemsara}%\thanks{sven.boenigk@fu-berlin.de}}
\author[1,2,3]{Isabel Naarmann-de Vries}%\thanks{ngehring@uni-koeln.de}}



\affil[1]{Klaus Tschira Institute for Integrative Computational Cardiology, University  Heidelberg, 69120 Heidelberg, Germany}
\affil[2]{Department of Internal Medicine III (Cardiology, Angiology, and Pneumology), University Hospital Heidelberg, 69120 Heidelberg, Germany}
\affil[3]{German Centre for Cardiovascular Research (DZHK)-Partner Site Heidelberg/Mannheim, 69120 Heidelberg, Germany}

\maketitle

\begin{abstract} % abstract
to be written
\end{abstract}

\keywords{Bayesian, 10X Genomics, Cell barcode assignment, Nonsense-mediated mRNA decay (NMD)}

\section*{INTRODUCTION}
Chemical modifications on DNA and histones, also known as epigenetics marks, strongly impact gene expression during cell differentiation and in several other biological programs. In the 1970s, it was recognized that RNA is also subjected to extensive covalent modification, and studies in the late 1980s revealed the widespread deamination of bases (termed RNA editing), which can lead to recoding if it occurs within coding sequences. Impressive development in the RNA modification field occurred during the past eight years, with the discovery of an extensive layer of base modifications in mRNAs. These can influence gene expression and have been already shown to be involved in primary cellular programs such as stem cell differentiation, response to stress, and the circadian clock. The study of RNA modifications and their effects is now referred to as epitranscriptomics, and it reveals striking similarities to what is known for epigenomics.
To date thirteen distinct modifications have been identified on mRNA transcripts \citep{Anreiter2021}. These modifications are catalyzed by a variety of dedicated enzymes and can be divided into two classes: modifications of cap-adjacent nucleotides and internal modifications. 

In contrast to the m7G cap, the impact of internal modifications on gene regulation has been less studied apart from RNA editing, which is mediated by RNA deaminases (e.g. the ADAR family). The most widespread internal mRNA modification is N6-methyladenosine (m6A). By modulating the processing of mRNA, m6A can regulate a wide range of physiological processes and its alteration has been linked to several diseases \cite{Roignant2017,Zaccara2019,Shi2019}. The modification is catalyzed co-transcriptionally by a Mega-Dalton methyltransferase complex, which includes the heterodimer METTL3-METTL14 and other associated subunits \cite{GarciasMorales2021}. This modification is reversible since two proteins of the AlkB-family demethylases can remove m6A from mRNA transcripts \citep{Jia2011,Zheng2013}. In mammals, m6A preferentially localizes within long internal exons and at the beginning of terminal exons at so-called DRACH motif (D = A/G/U, R = A/G, H = A/C/U) sites \citep{Dominissini2012,Meyer2012,Ke2015}. Once deposited, m6A is recognized by several reader proteins that can affect the fate of mRNA transcripts in nearly every step of the mRNA life cycle, which includes alternative splicing \citep{Adhikari2016,Roundtree2017}. The best-described readers are the YTH domain family of proteins that decode the signal and mediate m6A functions. By affecting RNA structure, m6A can also indirectly influence the association of additional RNA-binding proteins (RBPs) and the assembly of larger messenger ribonucleoprotein particles (mRNPs).

Several approaches have been presented to map RNA modifications on RNA. Herein, we focus on mRNA modification site detection in general and on m6A in particular where antibody-based protocols (miCLIP), methylation-sensitive restriction enzyme assays (MazF) or transgenic approaches (TRIBE, DART) have been presented. All of the aforementioned approaches rely on high-throughput sequencing on the Illumina platform. This typically involves cDNA synthesis by reverse transcription and PCR-based library amplification. One recent addition to the tool is direct RNA single molecule sequencing on the Oxford Nanopore Technology platform. While or software workflow is able to deal with Illumina and Nanopore-based approaches, the latter is the principal topic of our methods article.

\section*{MATERIALS}

\subsection*{ONT direct RNA sequencing}
\begin{enumerate}
\item 500 ng polyA\textsuperscript{+} RNA isolated from total RNA e.g. with Oligotex mRNA kit (Qiagen) or Dynabeads oligo dT\textsubscript{25} beads (Thermo Fisher Scientific) or \emph{in vitro} transcriptome sample. Store RNA at -80 $^{\circ}$C and the mRNA purification kit as recommended by the manufacturer.

\item Nuclease-free water. Store at room temperature.

\item Direct RNA-sequencing kit (SQK-RNA002, Oxford Nanopore Technologies). Store at -20 $^{\circ}$C.

\item NEBNext Quick Ligation Reaction Buffer (New England Biolabs). Store at -20 $^{\circ}$C.

\item T4 DNA Ligase (New England Biolabs). Store at -20 $^{\circ}$C.

\item dNTP Mix (10 mM each). Store at -20 $^{\circ}$C.

\item SuperScript IV Reverse Transcriptase (Thermo Fisher Scientific). Store at -20 $^{\circ}$C.

\item Agencourt RNAClean XP beads (Beckman Coulter). Store at 4 $^{\circ}$C.

\item 70 \% ethanol, freshly prepared. 

\item Qubit dsDNA HS assay kit and Qubit Fluorometer (Thermo Fisher Scientific).

\item Flow cell priming kit (EXP-FLP002, Oxford Nanopore Technologies). Store at -20 $^{\circ}$C.

\item Thermocycler.

\item Gentle rotator mixer.

\item Magnetic stand for 1.5 ml tubes.

\item 1.5 ml DNA LoBind tubes (Eppendorf), 0.2  ml PCR tubes.

\item MinION or GridION sequencing device and MinION R9.4.1 Flow cells (FLO-MIN106D, Oxford Nanopore Technologies). Store Flow cells at 4 $^{\circ}$C.
\end{enumerate}

\subsection*{Preparation of an \emph{in vitro} transcriptome sample}
\begin{enumerate}
\item 100 ng polyA\textsuperscript{+} RNA isolated from total RNA e.g. with Oligotex mRNA kit (Qiagen) or Dynabeads oligo dT\textsubscript{25} beads (Thermo Fisher Scientific). Store RNA at -80 $^{\circ}$C and the mRNA purification kit as recommended by the manufacturer

\item 10 $\mu$M oligo(dT)-VN RT primer. TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTVN. Store at -20 $^{\circ}$C.

\item 20 $\mu$M template switching oligo (TSO). ACTCTAATACGACTCACTATAGGGAGAGGGCrGrG+G. Store at -20 $^{\circ}$C.

\item 10 $\mu$M T7 extension primer. GCTCTAATACGACTCACTATAGG. Store at -20 $^{\circ}$C.

\item Nuclease-free water. Store at room temperature.

\item dNTP Mix (10 mM each). Store at -20 $^{\circ}$C.

\item Template Switching RT Enzyme Mix (New England Biolabs). Store at -20 $^{\circ}$C.

\item Q5 Hot Start High-Fidelity 2X Master Mix (New England Biolabs). Store at -20 $^{\circ}$C.

\item RNase H (5,000 U/ml) (New England Biolabs). Store at -20 $^{\circ}$C.

\item NucleoSpin Gel and PCR Cleanâ€‘up, Mini kit for gel extraction and PCR clean up (Macherey-Nagel) or equivalent. Store at room temperature.

\item MEGAscript T7 transcription kit (Thermo Fisher Scientific). Store at -20 $^{\circ}$C.

\item RNA Clean \& Concentrator-25 kit (Zymo Research). Store at room temperature.

\item Thermocycler.

\item Table top centrifuge for 1.5 ml tubes.

\item Nanodrop spectrophotometer or equivalent.

\item 0.2  ml PCR tubes, 1.5 ml DNA LoBind tubes (Eppendorf).

\end{enumerate}

\subsection*{Hardware requirements}
All analyses have been performed/tested on two alternative hardware systems:
a standard Linux desktop computer or an Apple iMac (Retina 5K, ultimo 2014).
The workflow requires a multi-core processor system with minimal main memory of 16GB RAM
and several GBs of free disk space (depending on data set size).

\subsection*{Software dependencies and installation}
Our analysis workflow has few requirements, which are detailed in Table \ref{tab:software}. Specifically, to execute our workflow, the following prerequisites are necessary: a BASH shell, a JAVA runtime environment, a working PERL and R installation. Additional i.e. non-standard software to process and map Nanopore reads (bedtools, samtools and Minimap2) are obligatory, while the installation of a Nanopore read simulator (NanoSim) is optional and depends on your use case. Table \ref{tab:packages} lists some additional R packages, which are required to run the R code. Detailed instructions on how to setup are found under \url{https://github.com/dieterich-lab/MiMB_JACUSA2_chapter} 

\section*{METHODS}
Overview Figure 1
\subsection*{Nanopore direct RNA sequencing}
\begin{enumerate}
\item Adjust 500 ng polyA\textsuperscript{+} RNA to a total volume of 9 $\mu$l with nuclease-free water. Complete RT adapter ligation reaction (in 0.2 ml PCR tube) with 3 $\mu$l NEBNext Quick Ligation Reaction Buffer, 0.5 $\mu$l RNA CS (RCS, from SQK-RNA002), 1 $\mu$l RT-Adapter (RTA, from SQK-RNA002) and 1.5 $\mu$l T4 DNA Ligase. Incubate 10 min at room temperature.

\item Prepare reverse transcription master mix on ice during ligation: 9 $\mu$l nuclease-free water, 2 $\mu$l 10 mM dNTPs, 8 $\mu$l 5x SuperScript IV first strand buffer, 4 $\mu$l 0.1 mM DTT.

\item Add the reverse transcription master mix to the ligation reaction and mix by pipetting. Add 2 $\mu$l SuperScript IV reverse transcriptase and mix by pipetting. Incubate in a thermocycler with the following protocol: 50 min at 50 $^{\circ}$C, 10 min at 70 $^{\circ}$C, cool down to 4 $^{\circ}$C.

\item Let the Agencourt RNAClean XP beads come to room temperature during reverse transcription. Carefully resuspend beads before use. Transfer reaction to a 1.5 ml DNA LoBind tube and mix with 72 $\mu$l Agencourt RNAClean XP beads. Incubate 5 min at room temperature on a gentle rotator mixer.

\item Collect beads on a magnetic stand and remove supernatant. Wash pelleted beads two times (30 sec) with 200 $\mu$l freshly prepared 70 \% ethanol. Remove supernatant. Spin sample down and place on magnet again. Remove any residual ethanol. 

\item Resuspend beads in 20 $\mu$l nuclease-free water by gentle flicking and incubate 5 min at room temperature on a gentle rotator mixer. Collect beads on a magnetic stand and transfer 20 $\mu$l eluate in a fresh 1.5 ml DNA LoBind tube.

\item For ligation of the RMX adapter, add the following to 20 $\mu$l eluate: 8 $\mu$l NEBNext Quick Ligation Reaction Buffer, 6 $\mu$l RMX (from SQK-RNA002), 3 $\mu$l nuclease-free water, 3 $\mu$l T4 DNA Ligase. Mix by pipetting and incubate 10 min at room temperature.

\item Add 40 $\mu$l carefully resuspended Agencourt RNAClean XP beads to the reaction and mix by pipetting. Incubate 5 min at room temperature on a gentle rotator mixer.

\item Collect beads on a magnetic stand and remove supernatant. Wash pelleted beads two times with 150 $\mu$l wash buffer (WSB, from SQK-RNA002). Resuspend beads by flicking, spin down and return to magnetic stand. Remove supernatant from pelleted beads.

\item Resuspend beads in 21 $\mu$l elution buffer (EB, from SQK-RNA002) by gentle flicking and incubate 5 min at room temperature on a gentle rotator mixer. Pellet beads on a magnetic stand and transfer 21 $\mu$l eluate in a fresh 1.5 ml DNA LoBind tube.

\item Quantify 1 $\mu$l of the library on a Qubit fluorometer with the Qubit dsDNA HS kit according to the manufacturerers protocol. Concentration should be usually in the range of 5 - 10 ng/$\mu$l.

\item Insert MinION R9.4.1 Flow cell in the MinION or GridION sequencing device and perform Flow cell check in the MinKNOW software. For successful sequencing of mammalian polyA\textsuperscript{+} RNA at least 1,000 available pores are recommended.

\item Prepare Priming Mix by adding 30 $\mu$l flush tether (FLT, from EXP-FLP002) to a vial of flush buffer (FB, from EXP-FLP002) and mix by pipetting. Open priming port. Remove air bubble from priming port by inserting the tip of a P1000 pipette into the priming port and slowly dialing up, until a small volume of storage buffer enters the pipette tip. Load 800 $\mu$l Priming Mix via the priming port and carefully avoid introduction of air bubbles. Close the priming port and wait for 5 min.

\item Mix 20 $\mu$l library with 17.5 $\mu$l nuclease-free water and 37.5 $\mu$l RNA running buffer (RRB, from SQK-RNA002) and mix by pipetting. Open the priming port and the sample port. Load 200 $\mu$l Priming Mix via the priming port. Mix library by pipetting just before loading and load dropwise via the sample port. Carefully avoid introduction of air bubbles. Close the sample port and the priming port.

\item Start sequencing for 48 to 72 h in the MinKNOW software. Choose direct RNA-sequencing kit and high-accuracy basecalling as parameters. We recommend to adjust the output filter to a minimum Q score of 7 (instead of 9).

\end{enumerate}

\subsection*{Preparation of an \emph{in vitro} transcriptome sample}
The \emph{in vitro} transcriptome sample is prepared based on a protocol published by Zhang \emph{et al.}~\cite{zhang2021systematic} with some modifications.

\begin{enumerate}
\item Adjust 100 ng polyA\textsuperscript{+} RNA to a total volume of 6 $\mu$l with nuclease-free water. Add 1 $\mu$l each of 10 $\mu$M oligo(dT)-VN RT primer and 10 mM dNTPs. Mix by pipetting and incubate in a thermocycler: 5 min at 75 $^{\circ}$C, 2 min at 42 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item Assemble 2.5 $\mu$l 4x template switching RT buffer, 0.5 $\mu$l 20 $\mu$M TSO, 1 $\mu$l 10x template switching RT enzyme mix and mix by pipetting. Combine with 6 $\mu$l RNA and incubate in a thermocycler: 90 min at 42 $^{\circ}$C, 10 min at 68 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item For Second strand synthesis add to First strand synthesis reaction: 50 $\mu$l Q5 Hot Start High-Fidelity 2X Master Mix, 5 $\mu$l RNase H, 2 $\mu$l 10 $\mu$M T7 extension primer, 33 $\mu$l nuclease-free water. Mix by pipetting and incubate in a thermocycler: 15 min at 37 $^{\circ}$C, 1 min at 95 $^{\circ}$C, 10 min at 65 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item Purify double stranded cDNA with NucleoSpin Gel and PCR Cleanâ€‘up kit according to the manufacturerers protocol and elute in 20 $\mu$l elution buffer. Determine concentration on a Nanodrop spectrophotometer. cDNA may be stored at -20 $^{\circ}$C.

\item Combine 8 $\mu$l cDNA for \emph{in vitro} transcription with 2 $\mu$l each of ATP, GTP, CTP, UTP, 10x reaction buffer and enzyme mix from the MEGAscript T7 transcription kit. Incubate 3 h at 37 $^{\circ}$C.

\item Digest template DNA by addition of 1 $\mu$l Turbo DNase. Mix by pipetting and incubate 15 min at 37 $^{\circ}$C.

\item Adjust reaction volume to 100 $\mu$l with nuclease-free water and clean up with RNA Clean \& Concentrator-25 kit according to the manufacturers protocol, using two volumes of adjusted RNA binding buffer (1:1 RNA binding buffer : ethanol). Elute RNA in 25 $\mu$l nuclease-free water. Determine RNA concentration on a Nanodrop spectrophotometer. Store at -80 $^{\circ}$C.

\end{enumerate}


\subsection*{Nanopore read processing}
\begin{enumerate}
	\item Following standard steps, base call the ionic current stored in FAST5 file using Guppy. The output is FASTQ files. For the IVT readout, we adopted real-time base calling with the integrated MinKNOW Guppy. Otherwise, Guppy base caller software could be used. The base caller requires the path to FAST5 files, the output path, and the config file or the flowcell/kit combination. More details can be found in \cite{Guppymanual}.
		\begin{spverbatim}
	$ guppy_basecaller --compress_fastq -i path_to_fast5 -s path_to_output --cpu_threads_per_caller 14 --num_callers 1 -c config_file.cfg
		\end{spverbatim}
	\item  Align reads to the transcriptome using Minimap2 software with the recommended setting for DirectRNA Sequencing (-ax map-ont). The output is a SAM file that should be converted into a compressed form as a BAM file using SAMtools command. The alignement requires the reference transcriptome/ genome. We used GRCh38 Ensembl annotations and FASTA file release version 96(\todo{Explain the rest of considered parameters}) . Check the Minimap2 manual for more details \cite{Minimap2manual}.
	\begin{spverbatim}
	$ minimap2 --secondary=no -ax map-ont -uf -k14 reference.fasta Reads.fastq |samtools view -bS > mapping.bam	
	\end{spverbatim}			
	\item  JACUSA2 requires sorted and indexed BAM files. To sort and create a BAM file index use the following SAMtools commands.
	\begin{spverbatim}
	$ samtools sorte mapping.bam mapping.sorted.bam
	$ samtools index mapping.sorted.bam
	\end{spverbatim}
%%	\item Add the MD tag field to BAM files. This requires the reference sequence 'reference.fasta' used for the mapping. The MD tag field stores information on mismatched and deleted reference bases.
%	\begin{spverbatim}
%	$ samtools calmd -b mapping.sorted.bam reference.fasta > align_md.bam
%	\end{spverbatim}
\end{enumerate}	
\subsection*{Use Case 1: Comparison of wildtype and knock-out samples}
We used a published dataset of Hek293 cell line \cite{pratanwanich2021identification}. The benchmark is composed of two samples from two conditions: wild type cells (modified RNAs) and Mettl3 knockout cells (unmodified RNAs) with two replicates (2 and 3). The FASTQ files are preprocessed and mapped according to the steps described above.
 
Given the preprocessed mapped reads as input (BAM files) 'HEK293T-WT-rep2.bam, HEK293T-WT-rep3.bam representing the wild type replicates and HEK293T-KO-rep2.bam and  HEK293T-KO-rep3.bam as the control replicates, 
\begin{enumerate} 
	\item Identify read error profile: run JACUSA2 in paired samples mode (call-2). The method requires setting the BAM files of the paired conditions, the corresponding library information, and the output file. Plus, it allows filtering reads according to many parameters. Here, we consider all sites with read coverage $> 4$. The output consists of a read error profile where the format is a combination of BED6 with JACUSA2 call-2 specific columns and common info columns: "info", "filter", and "ref". Check JACUSA2 manual for more details on JACUSA2 filter and output options \cite{JACUSA2manual}.
	\begin{spverbatim}
	$ JACUSA2 2.0.0-RC22 call-2 -m 1 -q 1 -c 4 -p 10 -D -I -a D,Y -P1 FR-SECONDSTRAND -P2 FR-SECONDSTRAND -r WT_vs_KO_call2_result.out HEK293T-WT-rep2.bam, HEK293T-WT-rep3.bam	HEK293T-KO-rep2.bam, HEK293T-KO-rep3.bam
	\end{spverbatim}
	\item Preprocess JACUSA2 output: given the JACUSA2 output, select non-overlapping sites of homo-polymer regions (JACUSA filter: Y) and within a 5mer of a central nucleotide A flanked by 2 adjacent random nucleotides (NNANN). Selected sites are characterized by the insertion, deletion and mismatch scores, and the position number within the specific 5mer context. The 'README\_processing.sh' bash script performs the preprocessing step and produces a text file 'call2\_SitesExt2\_indel\_slim2.txt' containing tabular features of the selected sites and a separate file representing the 5mer bases 'checkMotif\_reformat.txt'. The path to outputs could be specified within the command.
	\begin{spverbatim} 
	$ bash README_processing.sh WT_vs_KO_RC22_call2_result.out hg38.genome GRCh38_96.fa path_to_output.
	\end{spverbatim}
	\item Extract 5mer features: rebuild the tabular features such that the scores: mismatch, insertion and deletion are represented for each position of the specific 5mer (NNANN). To do so, run the R script 'HEK293\_data\_prep.R'. This produces an R object named 'BigTable.rds', representing the matrix of Sites$\times$15 features corresponding to the mismatch, insertion and deletion scores for the observed site and its two flanking positions. Be aware to precise the path to outputs that contains already the preprocessed data and provide the sample's name as a label of the analysis.
	\begin{spverbatim} 
	$ Rscript HEK293_data_prep.R path_to_output WT_vs_KO_RC22_call2_result.out
	\end{spverbatim}
	\item Extract m6A modification pattern: given the matrix of Sites$\times$Features, the next step is to learn a model representing the m6A modification pattern. To this end, the conventional non-negative matrix factorization (NMF) analysis is suggested \cite{lee1999learning}. Briefly, NMF factorizes a non-negative data matrix $X$ (here: $n$ sites and $m$ features) into two non-negative matrices as $X \approx WH$, such that $W$ is an $n\times k$ matrix containing basis vectors and $H$ is an $k \times m$ matrix containing coefficient vectors. The coefficient vectors and their combination can be viewed as a pattern for m6A modification. The R script 'HEK293\_data\_prep\_step2.R' allows generating patterns from a subset of the data related to previously reported m6A sites. Here, the unsupervised pattern training is based on $2401$ common reported m6A sites in \cite{koh2019atlas} {\todo{references}}. Based on the Silhouette and Cophenetic Correlation indices, we could identify an optimal factorization rank of 7. We then analyzed the identified patterns. According to the membership indicator of each site in matrix $W$, more than $80\%$ of m6A modification sites can be represented by five patterns (Patterns 2,3,4,6,7). The linear combination of these five patterns in fig (\ref{fig:WT_KO}A) highlights the importance of position 3 and eventually the implication of all scores.
	\begin{spverbatim} 
	$ Rscript HEK293_data_prep_step2.R path_to_output miCLIP_union.bed
	\end{spverbatim} 
	The 'miCLIP\_union.bed' file contains all m6A sites reported in \cite{koh2019atlas} (\todo{ reference}). 
\item Predict m6A modification: The additive linear combination of the coefficient vectors (patterns) with features can be used to predict m6A modification. We examine the ability of prediction on a subset of data of more than $1,98$ million sites with $22248$ miCLIP m6A validated sites {\todo{reference + description of categories}}. We opt for the linear combination of the five important patterns described in the previous section. The empirical Cumulative Distribution Function (eCDF) of the inferred scores shows a significant difference between the different miCLIP m6A categories and the unmodified sites (fig. \ref{fig:WT_KO}B). As the number of negative samples is much larger than the number of positive samples, we particularly recommend investigating the Positive Predictive Value (PPV) of the predictions. Here, (fig. \ref{fig:WT_KO}C) shows a moderate PPV that increases with the cut-off.  The R script 'HEK293\_data\_prep\_step3.R' allows generating the scores, eCDF probabilities of modification, and the corresponding eCDF and PPV plots.
	\begin{spverbatim} 
$ Rscript HEK293_data_prep_step3.R path_to_output miCLIP_union.bed
\end{spverbatim} 
	
\end{enumerate}

\subsection*{Use Case 2: Comparison of wildtype and IVT samples}
The second benchmarck is composed of wildtype HEK293 cell line from \cite{pratanwanich2021identification} and the synthetised IVT samples described in section 2.    
The analysis steps are similar to the first case. 
\begin{enumerate}
\item Identify read error profile: given the synthetised sequence, run JACUSA2 on paired conditions mode with the same parameters as the previously described case. 
\begin{verbatim}
$ JACUSA2 2.0.0-RC22 call-2 -m 1 -q 1 -c 4 -p 10 -D -I -a D,Y -P1 FR-SECONDSTRAND
-P2 FR-SECONDSTRAND -r WT_vs_realIVT_v202_call2_result.out HEK293T-WT-rep2.bam,HEK293T-WT-rep3.bam	HEK293T-IVT-rep1.bam,HEK293T-IVT-rep2.bam
\end{verbatim}
\item Preprocess JACUSA2 output: select the 5mer specific sites (NNANN) considering the Y filter as follows. 
\begin{verbatim} 
$ bash README_processing.sh WT_vs_IVT_RC22_call2_result.out hg38.genome GRCh38_96.fa path_to_output.
\end{verbatim}
\item Extract 5mer features using the following command:
\begin{verbatim} 
$ Rscript Code/HEK293_data_prep.R path_to_output WT_vs_IVT_RC22_call2_result.out
\end{verbatim}
\item Extract m6A modification patterns based on 2401 m6A sites (\todo{reference}). From the Silhouette and Cophenetic Correlation indices, we could identify an optimal factorization rank of 6 (fig. \ref{fig:WT_IVT}).
\begin{verbatim} 
$ Rscript HEK293_data_prep_step2.R path_to_output miCLIP_union_flat_exclude_Y_chromosome.bed
\end{verbatim} 
\item Predict m6A modifications: we examine the prediction of modification using the detected patterns on the test set of $22248$ m6A sites. So here, we plot the eCDF of pattern 6 scores by category (fig. \ref{fig:WT_IVT})
\begin{verbatim} 
$ Rscript HEK293_data_prep_step3.R path_to_output miCLIP_union_flat_exclude_Y_chromosome.bed
\end{verbatim} 

\end{enumerate}


\section*{NOTES}
\subsection*{Tips and Tricks}

\section*{ACKNOWLEDGMENTS}
  The authors would like to thank Etienne Boileau, Thiago Britto Borges, Tobias Jakobi for proof-reading and comments.
  The authors are grateful to Marek Franitza for running the experiments on the 10x platform and to Christian Becker for running ONT sequencing.
  This work was supported by Informatics for Life funded by the Klaus Tschira Foundation.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliography{Article} 

\newpage

\section*{FIGURE CAPTIONS}

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure1.pdf}
  \caption{\textbf{General outline of RNA modification detection by JACUSA2.} A key feature of our approach is that multiple replicates can be compared as shown on the left. Samples of interests where all modifications are present could be compared with either KO samples where the modification of interest is missing or IVT/simulated samples where all modifications are absent. Read stacks (in blue) are compared head-to-head as shown on the right. }
  \label{fig:graphicsummary}
      \end{figure}
\newpage

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure2.pdf}
  \caption{\textbf{Motivation of 5mer context for RNA modification mapping}. The nanopore covers 5 consecutive RNA residues. That is why we consider a 5mer context and derive 3 principal features for every position within a given 5 mer (15 features in total, with a central A residue in this example). We evaluate each feature set by previously learned patterns and compute a final score for modification site detection.}
  \label{fig:5mer}
      \end{figure}
\newpage

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure3.pdf}
  \caption{\textbf{Experimental and computational workflow}. tbd}
  \label{fig:workflow}
      \end{figure}
\newpage

\begin{figure}[h!]
	\includegraphics[width = 1\textwidth]{img/wt_ko.pdf}
	\caption{\textbf{Case 01. WT  versus KO.}  \textbf{A}: Barplots representing the linear combination of the top 5 patterns (y-axis) by position in the specific 5mer context (x-axis) and the score type: mismatch, deletion and insertion (resp. black, orange and blue). The 5 patterns (coefficient vectors: 2,3,4,6,7) are selected according to the predominant columns in matrix W.  \textbf{B}: Score distribution inferred from the combined patterns, stratified by the different categories of miCLIP validated sites and non miCLIP sites.  \textbf{C}: Number of predicted m6A sites (green) and Positive Predictive Value (PPV) of predicted m6A sites that overlap with miCLIP sites (orange). }
	\label{fig:WT_KO}
\end{figure}

\newpage
\begin{figure}[h!]
	\includegraphics[width = 1\textwidth]{img/wt_ivt.pdf}
	\caption{\textbf{Case 02. WT  versus IVT.}  \textbf{A}: Barplots representing the linear combination of the top 4 patterns (y-axis) by position in the specific 5mer context (x-axis) and the score type: mismatch, deletion and insertion (resp. black, orange and blue). The 4 patterns (coefficient vectors: 1,2,3,6) are selected according to the predominant columns in matrix W.  \textbf{B}: Score distribution inferred from the combined patterns, stratified by the different categories of miCLIP validated sites and non miCLIP sites.  \textbf{C}: Number of predicted m6A sites (green) and Positive Predictive Value (PPV) of predicted m6A sites that overlap with miCLIP sites (orange). }
	\label{fig:WT_IVT}
\end{figure}
\newpage
\section*{TABLE CAPTIONS}
\section*{TABLES}

\begin{table}[]
\label{tab:software}
\begin{tabular}{|p{1.5cm}|p{7cm}|p{6.5cm}|}
\hline 
            Software & Version & Description \\ 
            \hline \hline
            \texttt{Minimap2} & \url{https://github.com/lh3/minimap2} v2.22 or later & \url{https://lh3.github.io/minimap2/} \\ \hline
            \texttt{samtools} & \url{https://github.com/samtools/samtools} v1.12 or later & \url{http://samtools.github.io/} \\ \hline
	   \texttt{JAVA} & openjdk 11.0.12 2021-07-20 - JAVA 11 or later & OpenJDK Runtime Environment\\ \hline
	   \texttt{R} &  \url{https://www.r-project.org/} version 3.5.1 or later & The R Project for Statistical Computing \\ \hline
	   \texttt{PERL} &  \url{https://www.perl.org/} version 5.28.1 or later & Perl is a highly capable, feature-rich programming language \\ \hline
	   \texttt{BASH, sed, awk} &  should be part of your Linux distribution & Misc.\\ \hline
	   \texttt{bedtools} &  \url{https://github.com/arq5x/bedtools2} version 2.29.2 or later & Perl is a highly capable, feature-rich programming language \\ \hline	   
	   \texttt{NanoSim} &  \url{https://github.com/bcgsc/NanoSim} version 3.0.2 or later (optional) & NanoSim is a fast and scalable read simulator that captures the technology-specific features of ONT data\\ \hline	   
\end{tabular}
\caption{\textbf{Software dependencies}\label{tab:software} blubba}
\end{table}

\begin{table}[]
\label{tab:packages}
\begin{tabular}{|p{1.5cm}|p{7cm}|p{6.5cm}|}
\hline 
            R Packages & Version & Description \\ 
            \hline \hline
            \texttt{ggplot2} & \url{https://cran.r-project.org/web/packages/ggplot2/index.html} - ggplot2\_3.3.0 or later &  ggplot2 is a system for declaratively creating graphics, based on The Grammar of Graphics. \\ \hline
            \texttt{NMF} & \url{https://cran.r-project.org/web/packages/NMF/index.html} - NMF\_0.22.0 or later &  Provides a framework to perform Non-negative Matrix Factorization (NMF). \\ \hline
\end{tabular}
\caption{\textbf{R Package dependencies}\label{tab:software} blubba}
\end{table}



\end{document}
