\documentclass[times, 11pt, a4paper]{article}
\usepackage{authblk}
\usepackage{lineno}
\linenumbers

\usepackage[utf8]{inputenc} %unicode support
\usepackage{amsmath}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{todonotes}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{amssymb}
\DeclareCaptionLabelFormat{r-parens}{\textbf{#2}}
\captionsetup[subfigure]{font={bf,small}, skip=1pt, singlelinecheck=false, labelformat=r-parens}
\usepackage{graphicx}
\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage{hyperref}
\renewcommand{\refname}{REFERENCES}
\renewcommand\Authfont{\fontsize{10}{11}\selectfont}
\renewcommand\Affilfont{\fontsize{9}{11}\selectfont}
\begin{document}

\providecommand{\keywords}[1]{\textbf{Keywords:} #1}

\title{\Large Mapping of RNA modifications by direct Nanopore sequencing and JACUSA2}

\date{}
\author[1,2,3]{Christoph Dieterich\thanks{christoph.dieterich@uni-heidelberg.de}}
\author[1,2]{Amina Lemsara}%\thanks{sven.boenigk@fu-berlin.de}}
\author[1,2,3]{Isabel Naarmann-de Vries}%\thanks{ngehring@uni-koeln.de}}



\affil[1]{Klaus Tschira Institute for Integrative Computational Cardiology, University  Heidelberg, 69120 Heidelberg, Germany}
\affil[2]{Department of Internal Medicine III (Cardiology, Angiology, and Pneumology), University Hospital Heidelberg, 69120 Heidelberg, Germany}
\affil[3]{German Centre for Cardiovascular Research (DZHK)-Partner Site Heidelberg/Mannheim, 69120 Heidelberg, Germany}

\maketitle

\begin{abstract} % abstract
to be written
\end{abstract}

\keywords{Bayesian, 10X Genomics, Cell barcode assignment, Nonsense-mediated mRNA decay (NMD)}

\section*{INTRODUCTION}
Chemical modifications on DNA and histones, also known as epigenetics marks, strongly impact gene expression during cell differentiation and in several other biological programs. In the 1970s, it was recognized that RNA is also subjected to extensive covalent modification, and studies in the late 1980s revealed the widespread deamination of bases (termed RNA editing), which can lead to recoding if it occurs within coding sequences. Impressive development in the RNA modification field occurred during the past eight years, with the discovery of an extensive layer of base modifications in mRNAs. These can influence gene expression and have been already shown to be involved in primary cellular programs such as stem cell differentiation, response to stress, and the circadian clock. The study of RNA modifications and their effects is now referred to as epitranscriptomics, and it reveals striking similarities to what is known for epigenomics.
To date thirteen distinct modifications have been identified on mRNA transcripts \citep{Anreiter2021}. These modifications are catalyzed by a variety of dedicated enzymes and can be divided into two classes: modifications of cap-adjacent nucleotides and internal modifications. 

In contrast to the m7G cap, the impact of internal modifications on gene regulation has been less studied apart from RNA editing, which is mediated by RNA deaminases (e.g. the ADAR family). The most widespread internal mRNA modification is N6-methyladenosine (m6A). By modulating the processing of mRNA, m6A can regulate a wide range of physiological processes and its alteration has been linked to several diseases \cite{Roignant2017,Zaccara2019,Shi2019}. The modification is catalyzed co-transcriptionally by a Mega-Dalton methyltransferase complex, which includes the heterodimer METTL3-METTL14 and other associated subunits \cite{GarciasMorales2021}. This modification is reversible since two proteins of the AlkB-family demethylases can remove m6A from mRNA transcripts \citep{Jia2011,Zheng2013}. In mammals, m6A preferentially localizes within long internal exons and at the beginning of terminal exons at so-called DRACH motif (D = A/G/U, R = A/G, H = A/C/U) sites \citep{Dominissini2012,Meyer2012,Ke2015}. Once deposited, m6A is recognized by several reader proteins that can affect the fate of mRNA transcripts in nearly every step of the mRNA life cycle, which includes alternative splicing \citep{Adhikari2016,Roundtree2017}. The best-described readers are the YTH domain family of proteins that decode the signal and mediate m6A functions. By affecting RNA structure, m6A can also indirectly influence the association of additional RNA-binding proteins (RBPs) and the assembly of larger messenger ribonucleoprotein particles (mRNPs).

Several approaches have been presented to map RNA modifications on RNA. Herein, we focus on mRNA modification site detection in general and on m6A in particular where antibody-based protocols (miCLIP), methylation-sensitive restriction enzyme assays (MazF) or transgenic approaches (TRIBE, DART) have been presented. All of the aforementioned approaches rely on high-throughput sequencing on the Illumina platform. This typically involves cDNA synthesis by reverse transcription and PCR-based library amplification. One recent addition to the tool is direct RNA single molecule sequencing on the Oxford Nanopore Technology platform. While or software workflow is able to deal with Illumina and Nanopore-based approaches, the latter is the principal topic of our methods article.

\section*{MATERIALS}

\subsection*{ONT direct RNA sequencing}
ScNapBar offers two run modes. The first one uses cell barcode and UMI information without any additional modeling aspect. The second one introduces a probabilistic model, which performs very well in cases of low sequencing saturation (i.e. UMI coverage in Illumina data).

\subsection*{Software dependencies and installation}
The UMI approach requires a matching cell barcode and UMI tag and was first developed in Sicelore . Any cell barcode predictions that are supported by the presence of both, barcode and UMI alignment, are very reliable. We performed an \emph{in silico} benchmark of cell barcode assignment when both, cell barcode and UMI,  are found in the Nanopore read. We observed an average specificity of 99.9\% (ScNapBar) and 99.8\% (Sicelore) over 100 averaged simulation runs (Fig. \ref{fig:umimetrics}a). As expected, sensitivity heavily depends on Illumina sequencing saturation (Fig. \ref{fig:umimetrics}a). As the UMI approach relies on consistent genomic mappings for the Illumina and Nanopore reads, other challenges include: insufficient or inaccurate genome annotations causing wrong gene assignment; chimeric or super-long Nanopore reads assigned to multiple genes increase the risk of assigning a false UMI. %(Suppl. Fig. \ref{fig:simx}).

\section*{METHODS}
The current ecosystem of single-cell RNA-seq platforms is rapidly expanding, but robust solutions for single-cell and single-molecule full-length RNA sequencing are virtually absent. In our manuscript, we combined Oxford Nanopore single-molecule sequencing of 10x Genomics cDNA libraries and developed a novel software tool to arrive at single-cell, single-molecule, full cDNA length resolution. In contrast to , our Bayesian method for cell barcode assignment performs superior in situations of low sequencing saturation.
%This is related to issue 1
Even in the light of expected improvements of ONT sequencing error rates, ScNapBar offers improved performance the aforementioned use case (see Suppl. Fig. \ref{fig:lowerrormetrics}).  In summary, we could track in a well-controlled setting, i.e. by using GFP labeled cells and strong transcriptome pertubations, full-length transcript information at a single-cell level. We have identified differential RNA splicing linked to NMD pathway activity across our cell population. Our high-throughput full-length RNA sequencing solution is a necessary step forward towards studying the complex life of mRNA on single-cell level. This opens up unprecedented opportunities in low saturation settings such as multiplexed CRISPR-based screens.

\section*{NOTES}
\subsection*{Single cell samples preparation and experiment}

We performed an experiment using two different Flp-In-T-REx-293 cell lines: the wild type cell line with stably integrated FLAG-emGFP and a \textit{SMG7} knockout (KO) cell line (generated and established in ). Wild type cells (GFP+) were transfected with siRNA against Luciferase and the \textit{SMG7} KO cells (GFP-) were transfected with an siRNA against \textit{SMG6}. Two days after siRNA transfection, we mixed both cell types at a 1:1 ratio with a target of 2,000 cells in total.
cDNA was prepared according to the 10x Genomics Chromium Single Cell 3' Reagent Kit User Guide (v3  Chemistry) from the pool of $\approx$2.000 cells with a yield of 1.68 $\mu$g cDNA ( 42ng/$\mu$l concentration , 40$\mu$L volume) and a fragment size of 1.5kbp. At this point, all cDNA fragments carry the 16nt cellular barcode and 12nt UMI at the Poly (dT) end. 25\% (10 $\mu$L) of this cDNA solution continued with the original 10x Genomics protocol to create an Illumina 3' mRNA library with P5 and P7 Illumina adapters, and paired-end reads of this library will present cellular barcode and UMI with the first read and a 90nt second read containing 3' mRNA sequence. Further on, we produced two ONT libraries with 200ng each of the same cDNA  (0.2 pmoles) with the ONT Direct cDNA Sequencing Kit SQK-DCS109 protocol according to the manufactures' standard procedures. We sequenced each library on one GridIon flowcell (FLO-MIN106D R9 Version / R9.4.1) creating reads of full cDNA length that contained the same compositions of $\approx$2000 cellular barcodes as the Illumina data (based on the same cells) but a different composion of UMIs (different transcripts).

%A cDNA library was prepared according to the 10x Genomics Chromium Single Cell 3' Reagent Kit User Guide (v3 Chemistry) from the pool of cells. The final libraries contain the P5 and P7 primers. The P5 read contains 21-nt adaptor sequence, 16-nt cellular barcode, 12-nt UMI, and polyA-tail, followed by cDNA sequences.

\subsection*{Illumina reads processing and identification of cellular barcodes}
We used 10X Genomics Cell Ranger 3.1 (\url{https://github.com/10XGenomics/cellranger}) %\citep{10XGenomics2020}
 to map the Illumina reads onto the reference genome. In our NMD dataset, the DNA sequences of luciferase were appended to the reference genome, and therefore the GFP+ cells can be called from Cell Ranger. Cell Ranger also corrects the sequencing errors in the barcode and unique molecular identifier (UMI) sequences. Cell Ranger estimates the number of cells using a Good-Turing frequency estimation model (\url{https://support.10xgenomics.com}), %\citep{10XGenomics2020a}
 and characterized the identified barcodes into the cell-associated and background-associated barcodes. We used the cell-associated barcode sequences as the cellular barcode whitelist in the following analyses. Our CellRanger analysis estimated 2,052 sequenced cells (Suppl. Table \ref{tab:crs}). The read counts per cell of the estimated cell barcodes are shown in Suppl. Fig. \ref{fig:seqstats}. 

\section*{ACKNOWLEDGMENTS}
  The authors would like to thank Etienne Boileau, Thiago Britto Borges, Tobias Jakobi for proof-reading and comments.
  The authors are grateful to Marek Franitza for running the experiments on the 10x platform and to Christian Becker for running ONT sequencing.
  This work was supported by Informatics for Life funded by the Klaus Tschira Foundation.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliography{Article} 

\newpage

\section*{FIGURE CAPTIONS}

  \begin{figure}[h!]
    %\includegraphics[width = 1\textwidth]{figs/graphicsummary.pdf}
  \caption{\textbf{Combined Single-cell Illumina and Nanopore sequencing strategy.} GFP+/- cells are pooled and sequenced on the Illumina and Nanopore platform. The Nanopore platform generates long cDNA sequencing read that are used in barcode calling and estimating read error parameters. The Illumina data are used to estimate the total number of cells in sequencing and the represented cell barcodes. The simulated data are then used to parameterize a Bayesian model of barcode alignment features to discriminate correct vs. false barcode assignments. This model is then used on the real data to assign cell barcodes to Nanopore reads. The GFP label and known NMD transcripts can be used to validate this assignment. }
  \label{fig:graphicsummary}
      \end{figure}

\begin{figure}[h!]
     \centering
     \begin{subfigure}[t]{0.65\textwidth}
        % \caption{}
         \centering
         %\includegraphics[height=5cm]{figs/umi-t.pdf}
     \end{subfigure}
     \hfill
     \begin{subfigure}[t]{0.3\textwidth}
        % \caption{}
         \centering
         %\includegraphics[height=4.9cm]{figs/fc12-roc.pdf}
     %   \label{fig:score_metrics}
      \end{subfigure}
        \caption{\textbf{Sensitivity and specificity of ScNapBar and Sicelore on 100 Illumina libraries with different levels of saturation.} (a) Barcode assignment with UMI matches. (b) Barcode assignment without UMI matches (ScNapBar score \textgreater 50). (c) Benchmark of the specificity and sensitivity of the Illumina library with 100\% saturation. We compared the barcode assignments with ScNapBar score \textgreater 1-99, and the assignments from Sicelore with UMI support are roughly equivalent to the ScNapBar score \textgreater 90. }
         \label{fig:umimetrics}
\end{figure}
    
\begin{figure}[h!]
     \centering
     \begin{subfigure}[t]{0.35\textwidth}
         \caption{}
         \label{fig:cpu-rt}
         \centering
         %\includegraphics[width=\textwidth]{figs/cpu-rt.pdf}
     \end{subfigure}
     \hfill
     \begin{subfigure}[t]{0.55\textwidth}
         \caption{}
         \label{fig:ctime-sis}
         \centering
         %\includegraphics[width=\textwidth]{figs/cpu-sis.pdf}
     \end{subfigure}
        \caption{\textbf{Sicelore and ScNapBar CPU time comparison.} (a) ScNapBar CPU time depends on the number of whitelist barcodes (allowing an edit distance of  \textgreater 2 and and offset of up to 4bp between adapter and barcode). Gray area represents the standard deviation for 10 runs. (b) Comparison of ScNapBar and Sicelore CPU times. Benchmark was measured using one million barcode sequences and 2,052 barcodes in the whitelist. }
        \label{fig:runtime}
\end{figure}

\begin{figure}[h!]
     %\centering
     %\begin{subfigure}[b]{0.45\textwidth}
      %   \caption{}
         \centering
         %\includegraphics[width=\textwidth]{revision/figs/fc12-realsteps.pdf}
         \label{fig:fc1realsteps}
     %\end{subfigure}
     %\hfill
        \caption{\textbf{Number of the Nanopore reads identified by ScNapBar and Sicelore at each processing step.}  We inspected each processing step on real data (low lllumina saturation of 11.3\%). The first two steps are identical for both workflows. Total Reads: Number of input reads, Aligned to genome: Number of reads aligned to genome. The next three steps are workflow-specific: Aligned to adapter: Number of reads with identified adapter sequence, Aligned to barcode: Number of reads with aligned barcode sequence, Assigned to barcode: Number of predictions by each workflow. The last step is a validation of the previous assignment step after additional Illumina sequencing, which increases the Illumina saturation to 52\%, and using UMI matches, see main text.}
        \label{fig:realsteps}
\end{figure}

\section*{TABLE CAPTIONS}
\section*{TABLES}


\end{document}
