\documentclass[times, 11pt, a4paper]{article}
\usepackage{authblk}
\usepackage{lineno}
\linenumbers

\usepackage[utf8]{inputenc} %unicode support
\usepackage{amsmath}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{todonotes}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{amssymb}
\DeclareCaptionLabelFormat{r-parens}{\textbf{#2}}
\captionsetup[subfigure]{font={bf,small}, skip=1pt, singlelinecheck=false, labelformat=r-parens}
\usepackage{graphicx}
\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage{hyperref}
\renewcommand{\refname}{REFERENCES}
\renewcommand\Authfont{\fontsize{10}{11}\selectfont}
\renewcommand\Affilfont{\fontsize{9}{11}\selectfont}
\usepackage{spverbatim}
\begin{document}

\providecommand{\keywords}[1]{\textbf{Keywords:} #1}

\title{\Large Mapping of RNA modifications by direct Nanopore sequencing and JACUSA2}

\date{}
\author[1,2,3]{Christoph Dieterich\thanks{christoph.dieterich@uni-heidelberg.de}}
\author[1,2]{Amina Lemsara}%\thanks{sven.boenigk@fu-berlin.de}}
\author[1,2,3]{Isabel Naarmann-de Vries}%\thanks{ngehring@uni-koeln.de}}



\affil[1]{Klaus Tschira Institute for Integrative Computational Cardiology, University  Heidelberg, 69120 Heidelberg, Germany}
\affil[2]{Department of Internal Medicine III (Cardiology, Angiology, and Pneumology), University Hospital Heidelberg, 69120 Heidelberg, Germany}
\affil[3]{German Centre for Cardiovascular Research (DZHK)-Partner Site Heidelberg/Mannheim, 69120 Heidelberg, Germany}

\maketitle

\begin{abstract} % abstract
to be written
\end{abstract}

\keywords{Bayesian, 10X Genomics, Cell barcode assignment, Nonsense-mediated mRNA decay (NMD)}

\section*{INTRODUCTION}
Chemical modifications on DNA and histones, also known as epigenetics marks, strongly impact gene expression during cell differentiation and in several other biological programs. In the 1970s, it was recognized that RNA is also subjected to extensive covalent modification, and studies in the late 1980s revealed the widespread deamination of bases (termed RNA editing), which can lead to recoding if it occurs within coding sequences. Impressive development in the RNA modification field occurred during the past eight years, with the discovery of an extensive layer of base modifications in mRNAs. These can influence gene expression and have been already shown to be involved in primary cellular programs such as stem cell differentiation, response to stress, and the circadian clock. The study of RNA modifications and their effects is now referred to as epitranscriptomics, and it reveals striking similarities to what is known for epigenomics.
To date thirteen distinct modifications have been identified on mRNA transcripts \citep{Anreiter2021}. These modifications are catalyzed by a variety of dedicated enzymes and can be divided into two classes: modifications of cap-adjacent nucleotides and internal modifications. 

In contrast to the m7G cap, the impact of internal modifications on gene regulation has been less studied apart from RNA editing, which is mediated by RNA deaminases (e.g. the ADAR family). The most widespread internal mRNA modification is N6-methyladenosine (m6A). By modulating the processing of mRNA, m6A can regulate a wide range of physiological processes and its alteration has been linked to several diseases \cite{Roignant2017,Zaccara2019,Shi2019}. The modification is catalyzed co-transcriptionally by a Mega-Dalton methyltransferase complex, which includes the heterodimer METTL3-METTL14 and other associated subunits \cite{GarciasMorales2021}. This modification is reversible since two proteins of the AlkB-family demethylases can remove m6A from mRNA transcripts \citep{Jia2011,Zheng2013}. In mammals, m6A preferentially localizes within long internal exons and at the beginning of terminal exons at so-called DRACH motif (D = A/G/U, R = A/G, H = A/C/U) sites \citep{Dominissini2012,Meyer2012,Ke2015}. Once deposited, m6A is recognized by several reader proteins that can affect the fate of mRNA transcripts in nearly every step of the mRNA life cycle, which includes alternative splicing \citep{Adhikari2016,Roundtree2017}. The best-described readers are the YTH domain family of proteins that decode the signal and mediate m6A functions. By affecting RNA structure, m6A can also indirectly influence the association of additional RNA-binding proteins (RBPs) and the assembly of larger messenger ribonucleoprotein particles (mRNPs).

Several approaches have been presented to map RNA modifications on RNA. Herein, we focus on mRNA modification site detection in general and on m6A in particular where antibody-based protocols (miCLIP), methylation-sensitive restriction enzyme assays (MazF) or transgenic approaches (TRIBE, DART) have been presented. All of the aforementioned approaches rely on high-throughput sequencing on the Illumina platform. This typically involves cDNA synthesis by reverse transcription and PCR-based library amplification. One recent addition to the tool is direct RNA single molecule sequencing on the Oxford Nanopore Technology platform. While or software workflow is able to deal with Illumina and Nanopore-based approaches, the latter is the principal topic of our methods article.

\section*{MATERIALS}

\subsection*{ONT direct RNA sequencing}
\begin{enumerate}
\item 500 ng polyA\textsuperscript{+} RNA isolated from total RNA e.g. with Oligotex mRNA kit (Qiagen) or Dynabeads oligo dT\textsubscript{25} beads (Thermo Fisher Scientific) or \emph{in vitro} transcriptome sample. Store RNA at -80 $^{\circ}$C and the mRNA purification kit as recommended by the manufacturer.

\item Nuclease-free water. Store at room temperature.

\item Direct RNA-sequencing kit (SQK-RNA002, Oxford Nanopore Technologies). Store at -20 $^{\circ}$C.

\item NEBNext Quick Ligation Reaction Buffer (New England Biolabs). Store at -20 $^{\circ}$C.

\item T4 DNA Ligase (New England Biolabs). Store at -20 $^{\circ}$C.

\item dNTP Mix (10 mM each). Store at -20 $^{\circ}$C.

\item SuperScript IV Reverse Transcriptase (Thermo Fisher Scientific). Store at -20 $^{\circ}$C.

\item Agencourt RNAClean XP beads (Beckman Coulter). Store at 4 $^{\circ}$C.

\item 70 \% ethanol, freshly prepared. 

\item Qubit dsDNA HS assay kit and Qubit Fluorometer (Thermo Fisher Scientific).

\item Flow cell priming kit (EXP-FLP002, Oxford Nanopore Technologies). Store at -20 $^{\circ}$C.

\item Thermocycler.

\item Gentle rotator mixer.

\item Magnetic stand for 1.5 ml tubes.

\item 1.5 ml DNA LoBind tubes (Eppendorf), 0.2  ml PCR tubes.

\item MinION or GridION sequencing device and MinION R9.4.1 Flow cells (FLO-MIN106D, Oxford Nanopore Technologies). Store Flow cells at 4 $^{\circ}$C.
\end{enumerate}

\subsection*{Preparation of an \emph{in vitro} transcriptome sample}
\begin{enumerate}
\item 100 ng polyA\textsuperscript{+} RNA isolated from total RNA e.g. with Oligotex mRNA kit (Qiagen) or Dynabeads oligo dT\textsubscript{25} beads (Thermo Fisher Scientific). Store RNA at -80 $^{\circ}$C and the mRNA purification kit as recommended by the manufacturer

\item 10 $\mu$M oligo(dT)-VN RT primer. TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTVN. Store at -20 $^{\circ}$C.

\item 20 $\mu$M template switching oligo (TSO). ACTCTAATACGACTCACTATAGGGAGAGGGCrGrG+G. Store at -20 $^{\circ}$C.

\item 10 $\mu$M T7 extension primer. GCTCTAATACGACTCACTATAGG. Store at -20 $^{\circ}$C.

\item Nuclease-free water. Store at room temperature.

\item dNTP Mix (10 mM each). Store at -20 $^{\circ}$C.

\item Template Switching RT Enzyme Mix (New England Biolabs). Store at -20 $^{\circ}$C.

\item Q5 Hot Start High-Fidelity 2X Master Mix (New England Biolabs). Store at -20 $^{\circ}$C.

\item RNase H (5,000 U/ml) (New England Biolabs). Store at -20 $^{\circ}$C.

\item NucleoSpin Gel and PCR Cleanâ€‘up, Mini kit for gel extraction and PCR clean up (Macherey-Nagel) or equivalent. Store at room temperature.

\item MEGAscript T7 transcription kit (Thermo Fisher Scientific). Store at -20 $^{\circ}$C.

\item RNA Clean \& Concentrator-25 kit (Zymo Research). Store at room temperature.

\item Thermocycler.

\item Table top centrifuge for 1.5 ml tubes.

\item Nanodrop spectrophotometer or equivalent.

\item 0.2  ml PCR tubes, 1.5 ml DNA LoBind tubes (Eppendorf).

\end{enumerate}

\subsection*{Hardware requirements}
All analyses have been performed/tested on two alternative hardware systems:
a standard Linux desktop computer or an Apple iMac (Retina 5K, ultimo 2014).
The workflow requires a multi-core processor system with minimal main memory of 16GB RAM
and several GBs of free disk space (depending on data set size).

\subsection*{Software dependencies and installation}
Our analysis workflow has few requirements, which are detailed in Table \ref{tab:software}. Specifically, to execute our workflow, the following prerequisites are necessary: a BASH shell, a JAVA runtime environment, a working PERL and R installation. Additional i.e. non-standard software to process and map Nanopore reads (bedtools, samtools and Minimap2) are obligatory, while the installation of a Nanopore read simulator (NanoSim) is optional and depends on your use case. Table \ref{tab:packages} lists some additional R packages, which are required to run the R code. Detailed instructions on how to setup are found under \url{https://github.com/dieterich-lab/MiMB_JACUSA2_chapter} 

\section*{METHODS}
Overview Figure 1
\subsection*{Nanopore direct RNA sequencing}
\begin{enumerate}
\item Adjust 500 ng polyA\textsuperscript{+} RNA to a total volume of 9 $\mu$l with nuclease-free water. Complete RT adapter ligation reaction (in 0.2 ml PCR tube) with 3 $\mu$l NEBNext Quick Ligation Reaction Buffer, 0.5 $\mu$l RNA CS (RCS, from SQK-RNA002), 1 $\mu$l RT-Adapter (RTA, from SQK-RNA002) and 1.5 $\mu$l T4 DNA Ligase. Incubate 10 min at room temperature.

\item Prepare reverse transcription master mix on ice during ligation: 9 $\mu$l nuclease-free water, 2 $\mu$l 10 mM dNTPs, 8 $\mu$l 5x SuperScript IV first strand buffer, 4 $\mu$l 0.1 mM DTT.

\item Add the reverse transcription master mix to the ligation reaction and mix by pipetting. Add 2 $\mu$l SuperScript IV reverse transcriptase and mix by pipetting. Incubate in a thermocycler with the following protocol: 50 min at 50 $^{\circ}$C, 10 min at 70 $^{\circ}$C, cool down to 4 $^{\circ}$C.

\item Let the Agencourt RNAClean XP beads come to room temperature during reverse transcription. Carefully resuspend beads before use. Transfer reaction to a 1.5 ml DNA LoBind tube and mix with 72 $\mu$l Agencourt RNAClean XP beads. Incubate 5 min at room temperature on a gentle rotator mixer.

\item Collect beads on a magnetic stand and remove supernatant. Wash pelleted beads two times (30 sec) with 200 $\mu$l freshly prepared 70 \% ethanol. Remove supernatant. Spin sample down and place on magnet again. Remove any residual ethanol. 

\item Resuspend beads in 20 $\mu$l nuclease-free water by gentle flicking and incubate 5 min at room temperature on a gentle rotator mixer. Collect beads on a magnetic stand and transfer 20 $\mu$l eluate in a fresh 1.5 ml DNA LoBind tube.

\item For ligation of the RMX adapter, add the following to 20 $\mu$l eluate: 8 $\mu$l NEBNext Quick Ligation Reaction Buffer, 6 $\mu$l RMX (from SQK-RNA002), 3 $\mu$l nuclease-free water, 3 $\mu$l T4 DNA Ligase. Mix by pipetting and incubate 10 min at room temperature.

\item Add 40 $\mu$l carefully resuspended Agencourt RNAClean XP beads to the reaction and mix by pipetting. Incubate 5 min at room temperature on a gentle rotator mixer.

\item Collect beads on a magnetic stand and remove supernatant. Wash pelleted beads two times with 150 $\mu$l wash buffer (WSB, from SQK-RNA002). Resuspend beads by flicking, spin down and return to magnetic stand. Remove supernatant from pelleted beads.

\item Resuspend beads in 21 $\mu$l elution buffer (EB, from SQK-RNA002) by gentle flicking and incubate 5 min at room temperature on a gentle rotator mixer. Pellet beads on a magnetic stand and transfer 21 $\mu$l eluate in a fresh 1.5 ml DNA LoBind tube.

\item Quantify 1 $\mu$l of the library on a Qubit fluorometer with the Qubit dsDNA HS kit according to the manufacturerers protocol. Concentration should be usually in the range of 5 - 10 ng/$\mu$l.

\item Insert MinION R9.4.1 Flow cell in the MinION or GridION sequencing device and perform Flow cell check in the MinKNOW software. For successful sequencing of mammalian polyA\textsuperscript{+} RNA at least 1,000 available pores are recommended.

\item Prepare Priming Mix by adding 30 $\mu$l flush tether (FLT, from EXP-FLP002) to a vial of flush buffer (FB, from EXP-FLP002) and mix by pipetting. Open priming port. Remove air bubble from priming port by inserting the tip of a P1000 pipette into the priming port and slowly dialing up, until a small volume of storage buffer enters the pipette tip. Load 800 $\mu$l Priming Mix via the priming port and carefully avoid introduction of air bubbles. Close the priming port and wait for 5 min.

\item Mix 20 $\mu$l library with 17.5 $\mu$l nuclease-free water and 37.5 $\mu$l RNA running buffer (RRB, from SQK-RNA002) and mix by pipetting. Open the priming port and the sample port. Load 200 $\mu$l Priming Mix via the priming port. Mix library by pipetting just before loading and load dropwise via the sample port. Carefully avoid introduction of air bubbles. Close the sample port and the priming port.

\item Start sequencing for 48 to 72 h in the MinKNOW software. Choose direct RNA-sequencing kit and high-accuracy basecalling as parameters. We recommend to adjust the output filter to a minimum Q score of 7 (instead of 9).

\end{enumerate}

\subsection*{Preparation of an \emph{in vitro} transcriptome sample}
The \emph{in vitro} transcriptome sample is prepared based on a protocol published by Zhang \emph{et al.}~\cite{zhang2021systematic} with some modifications.

\begin{enumerate}
\item Adjust 100 ng polyA\textsuperscript{+} RNA to a total volume of 6 $\mu$l with nuclease-free water. Add 1 $\mu$l each of 10 $\mu$M oligo(dT)-VN RT primer and 10 mM dNTPs. Mix by pipetting and incubate in a thermocycler: 5 min at 75 $^{\circ}$C, 2 min at 42 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item Assemble 2.5 $\mu$l 4x template switching RT buffer, 0.5 $\mu$l 20 $\mu$M TSO, 1 $\mu$l 10x template switching RT enzyme mix and mix by pipetting. Combine with 6 $\mu$l RNA and incubate in a thermocycler: 90 min at 42 $^{\circ}$C, 10 min at 68 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item For Second strand synthesis add to First strand synthesis reaction: 50 $\mu$l Q5 Hot Start High-Fidelity 2X Master Mix, 5 $\mu$l RNase H, 2 $\mu$l 10 $\mu$M T7 extension primer, 33 $\mu$l nuclease-free water. Mix by pipetting and incubate in a thermocycler: 15 min at 37 $^{\circ}$C, 1 min at 95 $^{\circ}$C, 10 min at 65 $^{\circ}$C, cool to 4 $^{\circ}$C.

\item Purify double stranded cDNA with NucleoSpin Gel and PCR Cleanâ€‘up kit according to the manufacturerers protocol and elute in 20 $\mu$l elution buffer. Determine concentration on a Nanodrop spectrophotometer. cDNA may be stored at -20 $^{\circ}$C.

\item Combine 8 $\mu$l cDNA for \emph{in vitro} transcription with 2 $\mu$l each of ATP, GTP, CTP, UTP, 10x reaction buffer and enzyme mix from the MEGAscript T7 transcription kit. Incubate 3 h at 37 $^{\circ}$C.

\item Digest template DNA by addition of 1 $\mu$l Turbo DNase. Mix by pipetting and incubate 15 min at 37 $^{\circ}$C.

\item Adjust reaction volume to 100 $\mu$l with nuclease-free water and clean up with RNA Clean \& Concentrator-25 kit according to the manufacturers protocol, using two volumes of adjusted RNA binding buffer (1:1 RNA binding buffer : ethanol). Elute RNA in 25 $\mu$l nuclease-free water. Determine RNA concentration on a Nanodrop spectrophotometer. Store at -80 $^{\circ}$C.

\end{enumerate}


\subsection*{Nanopore read processing}
\begin{enumerate}
	\item Base call the ionic current signal stored in FAST5 file using Guppy. For the IVT readout, we adopted real-time base calling with the MinKNOW-embedded Guppy basecaller. Otherwise, Guppy basecaller software can be used; in this case, the basecaller requires the path to FAST5 files, the output folder, and the config file or the flowcell/kit combination. The output is FASTQ files that can be compressed using the option "--compress\_fastq".
		\begin{spverbatim}
	$ guppy_basecaller --compress_fastq -i path_to_fast5 -s path_to_output -c config_file.cfg --cpu_threads_per_caller 14 --num_callers 1
		\end{spverbatim}
	Set the number of threads "cpu\_threads\_per\_caller" and the number of parallel basecallers "num\_caller" according to your resources. Additional details can be found in \cite{Guppymanual}.
	\item  Align reads to the transcriptome using Minimap2 software. The output is a SAM file that has to be converted to a compressed form as BAM file using SAMtools command. The alignment requires the reference sequence. Here, we used GRCh38 Ensembl annotation and FASTA file release version 96. For Direct RNA Sequencing, it is recommended to use the default setting "-ax map-ont", "-uf" to force the alignment to the forward strand of the reference, and a small k-mer size "-k [=14]" to enhance sensitivity. We recommend to output primary alignments and up to "-N [=100]" top secondary alignments if the ratio of their chaining scores compared to the corresponding primary alignments is at least "-p [=1]" \todo{CD: not sure about minimap2 step specifically these 2 parameters -N and -p. need to be validated}. Customize the number of threads "-t" according to your resources. Check Minimap2 manual for more details \citep{Minimap2manual}.
	\begin{spverbatim}
	$ minimap2 -t 5 -ax map-ont -uf -k14 -p 1.0 -N 100 reference.fasta Reads.fastq |samtools view -bS > mapping.bam	
	\end{spverbatim}
	\item  Map RNA modifications using JACUSA2 software \citep{piechotta2021rna}. JACUSA2 rapidly detects RNA modifications based on a comparative strategy where the mapping features (mismatch, insertion and deletion) of a sample of interest is compared to a reference sequence (call-1) or against a sample without RNA modifications, e.g. a knock-out of an RNA modifying enzyme or an IVT (call-2). Moreover, it allows the integration of information from replicate experiments. Further details on how to use JACUSA2 is presented in the next section.
	
	Beforehand, JACUSA2 requires sorted and indexed BAM files. To sort a BAM file and create a BAM file index use the following SAMtools commands.
	\begin{spverbatim}
	$ samtools sorte mapping.bam mapping.sorted.bam
	$ samtools index mapping.sorted.bam
	\end{spverbatim}
%%	\item Add the MD tag field to BAM files. This requires the reference sequence 'reference.fasta' used for the mapping. The MD tag field stores information on mismatched and deleted reference bases.
%	\begin{spverbatim}
%	$ samtools calmd -b mapping.sorted.bam reference.fasta > align_md.bam
%	\end{spverbatim}
\end{enumerate}	
\subsection*{Use Case 1: Comparison of wild-type and knock-out samples}
The conventional way to detect RNA modifications using direct RNA sequencing is to compare a modified sample to an unmodified control sample. To assess the ability of JACUSA2 in this case, we used a published dataset of HEK293 cell lines to detect m6A modification \citep{pratanwanich2021identification}. The benchmark is composed of two sample sets from two conditions: wild-type cells (modified RNAs) and Mettl3 knockout cells (unmodified RNAs) with two replicates (2 and 3). The FASTQ files are preprocessed and mapped according to the steps described in the previous section. The analysis is validated against reported m6A sites in the three miCLIP-based studies \cite{boulias2019identification,koh2019atlas,kortel2021deep} (figure \ref{fig:miclip_sites}).
 
Given the preprocessed mapped reads as inputs (BAM files) 'HEK293T-WT-rep2.bam' and 'HEK293T-WT-rep3.bam' representing the wildtype replicates and 'HEK293T-KO-rep2.bam' and  'HEK293T-KO-rep3.bam' as the control replicates, 
\begin{enumerate} 
	\item Identify read error profile: run JACUSA2 in pairwise conditions mode (call-2). The method requires BAM files of the paired conditions, the corresponding library information "-P1" and "-P2", and the output file name "-r". In addition to the mismatch score, add "-D" and "-I" to output the deletion and insertion scores. JACUSA2 allows filtering reads according to many parameters. Here, we consider all sites with base calling quality "-q [$>1$]", mapping quality "-m [$>1$] and read coverage "-c [$>4$]". Plus, it provides a filter feature to improve sensitivity. Here, we consider the filter of sites within homopolymer regions "-a [=Y]". The output consists of a read error profile where the format is a combination of BED6 with JACUSA2 call-2 specific columns and common info columns: info, filter, and ref. Check JACUSA2 manual for more details on JACUSA2 filter and output options \citep{JACUSA2manual}. The number of threads can be customized via the parameter "-p".
	\begin{spverbatim}
	$ java -jar JACUSA2.jar call-2 -q 1 -m 1 -c 4 -p 10 -D -I -a Y -P1 FR-SECONDSTRAND -P2 FR-SECONDSTRAND -r WT_vs_KO_call2_result.out HEK293T-WT-rep2.bam, HEK293T-WT-rep3.bam	HEK293T-KO-rep2.bam, HEK293T-KO-rep3.bam
	\end{spverbatim}
	\item Preprocess JACUSA2 output: from JACUSA2 call-2 output, select all sites within 5-mer of a central nucleotide 'A' flanked by 2 random nucleotides (NNANN) and filter out sites of the homo-polymer regions (JACUSA filter: Y). 'README\_processing.sh' bash script performs the preprocessing step by providing the output of JAUSA2 call-2, the reference genome, and the output folder. The output of the preprocessing is a text file 'call2\_SitesExt2\_indel\_slim2.txt' containing tabular features of the selected sites. Eventually, sites are characterized by the main scores: mismatch, deletion, and insertion and additional information: reference base, strand and position number within the specific 5-mer context. 
	\begin{spverbatim} 
	$ bash README_processing.sh WT_vs_KO_call2_result.out hg38.genome GRCh38_96.fa path_to_output.
	\end{spverbatim}
	Then, using the R script 'HEK293\_data\_prep.R', rebuild the tabular features such that the observations are only sites with a reference base 'A'. Each site is characterized by 15 features corresponding to the mismatch, insertion and deletion scores for the observed site and its two flanking positions. The output is an R object 'BigTable.rds', representing the matrix of Sites$\times$15 features. Be aware to precise the path to outputs that contains already the preprocessed data and provide the sample's name as a label of the analysis.
	\begin{spverbatim} 
	$ Rscript HEK293_data_prep.R path_to_output WT_vs_KO_call2_result.out
	\end{spverbatim}
	\item Extract m6A modification pattern: given the matrix of Sites$\times$Features, the next step is to learn a model representing the m6A modification pattern. To this end, the conventional non-negative matrix factorization (NMF) analysis is suggested \citep{lee1999learning}. Briefly, NMF factorizes a non-negative data matrix $X$ (here: $n$ sites and $m$ features) into two non-negative matrices as $X \approx WH$, such that $W$ is an $n\times k$ matrix containing basis vectors and $H$ is an $k \times m$ matrix containing coefficient vectors. The coefficient vectors and their combination can be viewed as a pattern for m6A modification. The rank of factorization $k$ is a critical parameter that affects the performance substantially. We suggest to select the rank $k$ according to the method of \cite{frigyesi2008non} by looking at silhouette \citep{rousseeuw1987silhouettes} and cophenetic correlation \citep{brunet2004metagenes} indices. Accordingly, the performance indices are computed for different choices of rank ($k<n,m$) and compared to the performance of a random permutation of the original data. Subsequently, the chosen rank corresponds to the value with the largest difference between slopes of the original and the randomized data. Here, the unsupervised pattern training is based on the consensus set of $2,401$ m6A sites reported in the three miCLIP-based studies mentioned earlier. Based on the silhouette and cophenetic correlation indices, we could identify an optimal factorization rank of 7 (figure \ref{fig:WT_KO}A). We then analyzed the identified patterns. According to the membership indicator of each site in matrix $W$, more than $80\%$ of m6A modification sites can be represented by five patterns (Patterns 2,3,4,6,7) (figure \ref{fig:WT_KO}B). Interestingly, the linear combination of these five patterns in figure (\ref{fig:WT_KO}C) highlights the importance of position 3 and eventually the implication of all scores. 
	
	Use the R script 'HEK293\_data\_prep\_step2.R' to generate patterns and provide the set of modified sites as a ground truth. Here, the 'miCLIP\_union.bed' file contains the m6A sites from the three miCLIP-based studies. A miCLIP annotation, reflecting studies (hence, the consensus) wherein the modification is reported, is added to each site.
	
	\begin{spverbatim} 
	$ Rscript HEK293_data_prep_step2.R path_to_output miCLIP_union.bed
	\end{spverbatim} 
\item Predict m6A modifications: the additive linear combination of the coefficient vectors (patterns) with the 15 features can be used to predict m6A modification. We examine the ability of prediction on a subset of data of more than $1,98$ million sites with $22,248$ miCLIP m6A sites. We opt for the linear combination of the five important patterns described in step 3. The empirical Cumulative Distribution Function (eCDF) of the inferred scores shows clearly a significant difference between the different miCLIP m6A categories (miCLIP annotation) and the unmodified sites (figure \ref{fig:WT_KO}D). As the number of negative samples is much larger than the number of positive samples, we particularly recommend investigating the Positive Predictive Value (PPV) of the predictions. Here, figure \ref{fig:WT_KO}E shows a moderate PPV that increases with the cut-off.  

Use the R script 'HEK293\_data\_prep\_step3.R' to estimate scores and eCDF probability of modification from the selected patterns and to produce the corresponding PPV plot.
	\begin{spverbatim} 
$ Rscript HEK293_data_prep_step3.R path_to_output miCLIP_union.bed
\end{spverbatim} 
	
\end{enumerate}

\subsection*{Use Case 2: Comparison of wild-type and IVT samples}
An alternative way to detect RNA modification is to compare a modified sample to an \emph{in-vitro} (IVT) synthesized control sample. Therefore, we benchmark JACUSA2 on a sample set of wild-type HEK293 cell lines (modified sample) with two replicates (2 and 3) from \cite{pratanwanich2021identification} and a modification-free RNA synthesized sample (control sample).    
The analysis steps are similar to case 1. We evaluate the analysis against miCLIP m6A sites (figure \ref{fig:miclip_sites}).
\begin{enumerate}
\item Identify read error profile: we use JACUSA2 call-2 with the same parameters as the previously described case. The input BAM files (HEK293T-WT-rep2.bam, HEK293T-WT-rep3.bam) and (HEK293T-IVT-rep1.bam, HEK293T-IVT-rep2.bam) are associated to the wild-type and IVT replicate samples respectively.
\begin{spverbatim}
$ java -jar JACUSA2.jar call-2 -m 1 -q 1 -c 4 -p 10 -D -I -a D,Y -P1 FR-SECONDSTRAND
-P2 FR-SECONDSTRAND -r WT_vs_IVT_call2_result.out HEK293T-WT-rep2.bam, HEK293T-WT-rep3.bam	HEK293T-IVT-rep1.bam, HEK293T-IVT-rep2.bam
\end{spverbatim}
\item Preprocess JACUSA2 output: we select all sites within the specific 5-mer (NNANN) and we consider the Y filter that excludes sites within the homo-polymer regions. 
\begin{spverbatim} 
$ bash README_processing.sh WT_vs_IVT_call2_result.out hg38.genome GRCh38_96.fa path_to_output.
\end{spverbatim}
Then, we extract 5-mer features such that the selected sites are represented by the three scores: mismatch, deletion and insertion for the observed site and its two flanking positions. 
\begin{spverbatim} 
$ Rscript HEK293_data_prep.R path_to_output WT_vs_IVT_call2_result.out
\end{spverbatim}
\item Extract m6A modification pattern: using NMF factorization, we extract patterns from $1,905$ sites reported as modified in the three miCLIP-based studies. Based on the silhouette and cophenetic correlation indices, we could identify an optimal factorization rank of 6 (figure \ref{fig:WT_IVT}A). We determined the predominant factors from matrix $W$; accordingly, more than $80\%$ of m6A modification sites can be represented by four patterns (Patterns: 1,2,3,6) (figure \ref{fig:WT_IVT}B). In agreement with case 1, the linear combination of the four patterns confirms the importance of position 3 and the implication of all scores as shown in figure (\ref{fig:WT_IVT}C).
\begin{spverbatim} 
$ Rscript HEK293_data_prep_step2.R path_to_output miCLIP_union.bed
\end{spverbatim} 
\item Predict m6A modifications: we evaluate the prediction ability of the detected patterns on a test set of almost $1,52$ million sites where $17,021$ are miCLIP-m6A modified. We consider the linear combination of the four important patterns (1,2,3,6). Figure \ref{fig:WT_IVT}D shows the eCDF of the inferred scores. The difference between the cumulative distribution of non miCLIP sites and miCLIP sites can be nicely observed, while, the PPV plot shows a lower performance as compared to case 1 (figure \ref{fig:WT_IVT}E). The decrease in performance is likely explained by the absence of all modifications and not exclusively m6A in the control condition, which may induce noise to the score estimation by JACUSA2 call-2 \todo{CD:to be confirmed}.
\begin{spverbatim} 
$ Rscript HEK293_data_prep_step3.R path_to_output miCLIP_union.bed
\end{spverbatim} 

\end{enumerate}


\section*{NOTES}
\subsection*{Tips and Tricks}

\section*{ACKNOWLEDGMENTS}
  The authors would like to thank Etienne Boileau, Thiago Britto Borges, Tobias Jakobi for proof-reading and comments.
  The authors are grateful to Marek Franitza for running the experiments on the 10x platform and to Christian Becker for running ONT sequencing.
  This work was supported by Informatics for Life funded by the Klaus Tschira Foundation.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliography{Article} 

\newpage

\section*{FIGURE CAPTIONS}

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure1.pdf}
  \caption{\textbf{General outline of RNA modification detection by JACUSA2.} A key feature of our approach is that multiple replicates can be compared as shown on the left. Samples of interests where all modifications are present could be compared with either KO samples where the modification of interest is missing or IVT/simulated samples where all modifications are absent. Read stacks (in blue) are compared head-to-head as shown on the right. }
  \label{fig:graphicsummary}
      \end{figure}
\newpage

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure2.pdf}
  \caption{\textbf{Motivation of 5-mer context for RNA modification mapping}. The nanopore covers 5 consecutive RNA residues. That is why we consider a 5-mer context and derive 3 principal features for every position within a given 5-mer (15 features in total, with a central A residue in this example). We evaluate each feature set by previously learned patterns and compute a final score for modification site detection.}
  \label{fig:5mer}
      \end{figure}
\newpage

\begin{figure}[h!]
    \includegraphics[width = 1\textwidth]{Figure3.pdf}
  \caption{\textbf{Experimental and computational workflow}. tbd}
  \label{fig:workflow}
      \end{figure}
\newpage

\begin{figure}[h!]
	\includegraphics[width = 0.6\textwidth]{img/venn_diagramm.png}
	\caption{m6A sites reported in the three miCLIP-based studies: Boulias et al. [2019], Koh et al. [2019] and KÃ¶rtel et al. [2021].}
	\label{fig:miclip_sites}
\end{figure}
\newpage

\begin{figure}[h!]
	\includegraphics[width = 1\textwidth]{img/res_wt_ko.pdf}
	\caption{\textbf{Case 1. WT  versus KO.}  \textbf{A}: NMF rank selection. Comparison of cophenetic correlation and silhouette indices obtained from the NMF factorization of the original and the randomized data \textbf{B}: NMF result represented by the basis matrix $W$ and the coefficient matrix $H$. The matrix $H$ induces the RNA modification pattern. \textbf{B}: Barplots representing the linear combination of the top 5 patterns (y-axis) by the number of position in the specific 5-mer context (x-axis) and the score type: mismatch, deletion and insertion (resp. black, orange and blue). The 5 patterns (coefficient vectors: 2,3,4,6,7) are selected according to the predominant columns in matrix W.  \textbf{C}: Score distribution inferred from the combined patterns, stratified by the different categories of miCLIP validated sites and non miCLIP sites.  \textbf{D}: Number of predicted m6A sites (green) and Positive Predictive Value (PPV) of predicted m6A sites that overlap with miCLIP sites (orange). }
	\label{fig:WT_KO}
\end{figure}

\newpage
\begin{figure}[h!]
	\includegraphics[width = 1\textwidth]{img/res_wt_ivt.pdf}
	\caption{\textbf{Case 2. WT  versus IVT.}  \textbf{A}: NMF rank selection. Comparison of cophenetic correlation and silhouette indices resulting from the NMF factorization of the original and the randomized data \textbf{B}: NMF result represented by the basis matrix $W$ and the coefficient matrix $H$. The matrix $H$ induces the RNA modification pattern. \textbf{B}: Barplots representing the linear combination of the top 4 patterns (y-axis) by by the number of position in the specific 5-mer context (x-axis) and the score type: mismatch, deletion and insertion (resp. black, orange and blue). The 4 patterns (coefficient vectors: 1,2,3,6) are selected according to the predominant columns in matrix W.  \textbf{C}: Score distribution inferred from the combined patterns, stratified by the different categories of miCLIP validated sites and non miCLIP sites.  \textbf{D}: Number of predicted m6A sites (green) and Positive Predictive Value (PPV) of predicted m6A sites that overlap with miCLIP sites (orange). }
	\label{fig:WT_IVT}
\end{figure}
\newpage
\section*{TABLE CAPTIONS}
\section*{TABLES}

\begin{table}[]
\label{tab:software}
\begin{tabular}{|p{1.5cm}|p{7cm}|p{6.5cm}|}
\hline 
            Software & Version & Description \\ 
            \hline \hline
            \texttt{Minimap2} & \url{https://github.com/lh3/minimap2} v2.22 or later & \url{https://lh3.github.io/minimap2/} \\ \hline
            \texttt{samtools} & \url{https://github.com/samtools/samtools} v1.12 or later & \url{http://samtools.github.io/} \\ \hline
	   \texttt{JAVA} & openjdk 11.0.12 2021-07-20 - JAVA 11 or later & OpenJDK Runtime Environment\\ \hline
	   \texttt{R} &  \url{https://www.r-project.org/} version 3.5.1 or later & The R Project for Statistical Computing \\ \hline
	   \texttt{PERL} &  \url{https://www.perl.org/} version 5.28.1 or later & Perl is a highly capable, feature-rich programming language \\ \hline
	   \texttt{BASH, sed, awk} &  should be part of your Linux distribution & Misc.\\ \hline
	   \texttt{bedtools} &  \url{https://github.com/arq5x/bedtools2} version 2.29.2 or later & Perl is a highly capable, feature-rich programming language \\ \hline	   
	   \texttt{NanoSim} &  \url{https://github.com/bcgsc/NanoSim} version 3.0.2 or later (optional) & NanoSim is a fast and scalable read simulator that captures the technology-specific features of ONT data\\ \hline	   
\end{tabular}
\caption{\textbf{Software dependencies}\label{tab:software} blubba}
\end{table}

\begin{table}[]
\label{tab:packages}
\begin{tabular}{|p{1.5cm}|p{7cm}|p{6.5cm}|}
\hline 
            R Packages & Version & Description \\ 
            \hline \hline
            \texttt{ggplot2} & \url{https://cran.r-project.org/web/packages/ggplot2/index.html} - ggplot2\_3.3.0 or later &  ggplot2 is a system for declaratively creating graphics, based on The Grammar of Graphics. \\ \hline
            \texttt{NMF} & \url{https://cran.r-project.org/web/packages/NMF/index.html} - NMF\_0.22.0 or later &  Provides a framework to perform Non-negative Matrix Factorization (NMF). \\ \hline
\end{tabular}
\caption{\textbf{R Package dependencies}\label{tab:software} blubba}
\end{table}



\end{document}
